openerp/addons/base/ir/ir_translation.py

```diff
@@ -284,13 +284,26 @@ def _set_ids(self, cr, uid, name, tt, lang, ids, value, src=None):
         self._get_ids.clear_cache(self)
         self._get_source.clear_cache(self)
 
-        cr.execute('delete from ir_translation '
+        cr.execute('update ir_translation '
+                  'set value=%s '
+                  '  , src=%s '
+                  '  , state=%s '
                 'where lang=%s '
                     'and type=%s '
                     'and name=%s '
                     'and res_id IN %s',
-                (lang,tt,name,tuple(ids),))
-        for id in ids:
+                (value,src,'translated',lang,tt,name,tuple(ids),))
+
+        cr.execute('select res_id '
+                'from ir_translation '
+                'where lang=%s '
+                    'and type=%s '
+                    'and name=%s '
+                    'and res_id IN %s',
+                (lang,tt,name,tuple(ids)))
+        existing_ids = [x[0] for x in cr.fetchall()]
```

Comment:
You may combine this query with the former one by adding `RETURNING id` to it, and fetching its result.